onelogin_aws_cli/credentials.py

```diff
@@ -0,0 +1,111 @@
+"""
+Handles the saving and loading of username and password in a secure
+manner
+"""
+
+import getpass
+import keyring
+
+from onelogin_aws_cli.configuration import Section
+from typing import Optional
+
+
+class UserCredentials(object):
+    """
+    Class to encapsulate the handling of storing and retrieving user password
+    in OS-Independent system keychain.
+    """
+    SERVICE_NAME = "onelogin-aws-cli"
+
+    def __init__(self, username, config: Section):
+        self.username = username
+        self.configuration = config
+
+        # This is `None`, as the password should be be emitted from this class
+        # and should never be loaded from any other source outside this class
+        self.password = None
+
+    @property
+    def has_password(self) -> bool:
+        """
+        True if the class has a password.
+
+        :return:Whether we have set a password or not, yet
+        """
+        return (self.password is not None) and \
+               (self.password != "")
+
+    def load_credentials(self):
+        """
+        Load the username and password
+        """
+
+        self.load_username()
+        self.load_password()
+
+    def load_username(self):
+        """
+        Either load the username from configfile or prompt the user to supply
+        one interactively
+
+        :return: username
+        """
+
+        if not self.username:
+            # Try the configurationfile first
+            if 'username' in self.configuration:
+                username = self.configuration['username']
+            else:
+                username = input("Onelogin Username: ")
+            self.username = username
+
+    def load_password(self):
+        """
+        Load the password from keychain if we expect to be able to save the
+        password in a keychain, or prompt the user for a password through
+        stdin.
+
+        :return: user password
+        """
+
+        save_password = False
+
+        # Do we have a password?
+        if not self.has_password:
+            # Can we load the password from os keychain?
+            if self.configuration.can_save_password:
```

Comment:
I occurs to me that this might be simpler and more future proof if the config parameter was just something like auth_methods='...,...'  
and this function just calls the other methods specified in the string.

Reply:
I’m not following. Can you elaborate?

Reply:
I was imagining something like the following (which kind of mirrors the way I think ssh does this):

```
auth_methods = self.configuration.auth_methods.split(',')

for am in auth_methods:
   if hasattr(self, am):
      self.password = getattr(self, am)()
      if self.password:
        break
```

and then elsewhere

```
def password(self):
   getpass.getpass("Onelogin Password: ")
```

Just noodling around but it's not a big deal =).

Reply:
ooohhh gotcha

Reply:
Thinking further down that route, if we wanted to go that way, I'd say do a plugin architecture
ala https://packaging.python.org/guides/creating-and-discovering-plugins/

Then there's no need for core package changes. I think it would be a simple-ish PR.